bootstrap/kubeadm/internal/ignition/clc/clc.go

```diff
@@ -0,0 +1,163 @@
+	"sigs.k8s.io/cluster-api/bootstrap/kubeadm/internal/cloudinit"
+)
+
+const (
+	clcTemplate = `---
+systemd:
+  units:
+    - name: kubeadm.service
+      enabled: true
+      contents: |
+        [Unit]
+        Description=kubeadm
+        [Service]
+        # To not restart the unit when it exits, as it is expected.
+        Type=oneshot
+        ExecStart=/etc/kubeadm.sh
+        [Install]
+        WantedBy=multi-user.target
+
+storage:
+  files:{{ range .WriteFiles }}
+  - path: {{.Path}}
+    {{ if ne .Permissions "" -}}
+    mode: {{.Permissions}}
+    {{ end -}}
+    contents:
+      inline: |
+{{.Content | Indent 8}}
+{{- end }}
+  - path: /etc/kubeadm.sh
+    mode: 0700
+    contents:
+      inline: |
+        #!/bin/bash
+        set -e
+
+        {{- range .PreKubeadmCommands }}
+        {{ . }}
+        {{- end }}
+
+        {{ .KubeadmCommand }}
+        rm /etc/kubeadm.yml
+
+        {{- range .PostKubeadmCommands }}
+        {{ . }}
+        {{- end }}
+  - path: /etc/kubeadm.yml
+    mode: 0600
+    contents:
+      inline: |
+        ---
+{{ .KubeadmConfig | Indent 8 }}
+`
+)
+
+type render struct {
+	*cloudinit.BaseUserData
+
+	KubeadmConfig string
+}
+
+func defaultTemplateFuncMap() template.FuncMap {
+	return template.FuncMap{
+		"Indent": templateYAMLIndent,
+	}
+}
+
+func templateYAMLIndent(i int, input string) string {
+	split := strings.Split(input, "\n")
+	ident := "\n" + strings.Repeat(" ", i)
+	return strings.Repeat(" ", i) + strings.Join(split, ident)
+}
+
+func renderCLC(input *cloudinit.BaseUserData, kubeadmConfig string) ([]byte, error) {
+	t := template.Must(template.New("template").Funcs(defaultTemplateFuncMap()).Parse(clcTemplate))
+
+	data := render{
+		BaseUserData:  input,
+		KubeadmConfig: kubeadmConfig,
+	}
+
+	var out bytes.Buffer
+	if err := t.Execute(&out, data); err != nil {
+		return nil, errors.Wrapf(err, "failed to render template")
+	}
+
+	return out.Bytes(), nil
+}
+
+func Render(input *cloudinit.BaseUserData, clc *bootstrapv1.ContainerLinuxConfig, kubeadmConfig string) ([]byte, error) {
+	clcBytes, err := renderCLC(input, kubeadmConfig)
+	if err != nil {
+		return nil, errors.Wrapf(err, "rendering CLC configuration")
+	}
+
+	ign, err := clcToIgnition(clcBytes)
+	if err != nil {
+		return nil, errors.Wrapf(err, "converting generated CLC to Ignition")
+	}
+
+	if clc.AdditionalConfig != "" {
+		additionalIgn, err := clcToIgnition([]byte(clc.AdditionalConfig))
+		if err != nil {
+			return nil, errors.Wrapf(err, "converting additional CLC to Ignition")
+		}
+
+		ign = ignition.Append(ign, additionalIgn)
+	}
+
+	userData, err := json.Marshal(&ign)
+	if err != nil {
+		return nil, errors.Wrapf(err, "marshaling generated Ignition config into JSON")
+	}
+
+	return userData, nil
+}
+
+func clcToIgnition(data []byte) (ignitionTypes.Config, error) {
+	clc, ast, report := clct.Parse(data)
+
+	if len(report.Entries) > 0 || report.IsFatal() {
```

Comment:
The report can't be fatal unless it has entries.

If `len(report.Entries) > 0 && !report.IsFatal()`, there are warnings but no errors.  In that case we might want to continue but log the report contents somewhere (preferably somewhere user-accessible if that's at all possible).

Same comment below.

Reply:
True. Hmm, it seems container runtime webhooks do not expose any way to emit the warnings. I guess I could:
- Make controller emit a condition with warnings.
- Add a strict field to be able to control if we should fail on warnings.

Reply:
I added both, please have a look again :)